core/services/ocr2/plugins/ocr2keeper/integration_21_test.go

```diff
@@ -592,3 +728,292 @@ func getUpkeepIdFromTx21(t *testing.T, registry *iregistry21.IKeeperRegistryMast
+	registryOwner *bind.TransactOpts,
+	backend *backends.SimulatedBackend,
+	linkToken *link_token_interface.LinkToken,
+) registerAndFundFunc {
+	return func(t *testing.T, upkeepAddr common.Address, upkeepOwner *bind.TransactOpts, trigger uint8, config []byte) *big.Int {
+		// register the upkeep on the host registry contract
+		registrationTx, err := registry.RegisterUpkeep(
+			registryOwner,
+			upkeepAddr,
+			2_500_000,
+			upkeepOwner.From,
+			trigger,
+			[]byte{},
+			config,
+			[]byte{},
+		)
+		require.NoError(t, err)
+
+		backend.Commit()
+
+		receipt, err := backend.TransactionReceipt(testutils.Context(t), registrationTx.Hash())
+		require.NoError(t, err)
+
+		parsedLog, err := registry.ParseUpkeepRegistered(*receipt.Logs[0])
+		require.NoError(t, err)
+
+		upkeepID := parsedLog.Id
+
+		// Fund the upkeep
+		_, err = linkToken.Approve(upkeepOwner, registry.Address(), oneHunEth)
+		require.NoError(t, err)
+
+		_, err = registry.AddFunds(upkeepOwner, upkeepID, oneHunEth)
+		require.NoError(t, err)
+
+		backend.Commit()
+
+		return upkeepID
+	}
+}
+
+type feedLookupUpkeepController struct {
+	// address for dummy protocol
+	logSrcAddr common.Address
+	// dummy protocol is a log event source
+	protocol      *dummy_protocol_wrapper.DummyProtocol
+	protocolOwner *bind.TransactOpts
+	// log trigger listener contracts react to logs produced from protocol
+	count          int
+	upkeepIds      []*big.Int
+	addresses      []common.Address
+	contracts      []*log_triggered_streams_lookup_wrapper.LogTriggeredStreamsLookup
+	contractsOwner *bind.TransactOpts
+}
+
+func newFeedLookupUpkeepController(
+	backend *backends.SimulatedBackend,
+	protocolOwner *bind.TransactOpts,
+) (*feedLookupUpkeepController, error) {
+	addr, _, contract, err := dummy_protocol_wrapper.DeployDummyProtocol(protocolOwner, backend)
+	if err != nil {
+		return nil, err
+	}
+
+	backend.Commit()
+
+	return &feedLookupUpkeepController{
+		logSrcAddr:    addr,
+		protocol:      contract,
+		protocolOwner: protocolOwner,
+	}, nil
+}
+
+func (c *feedLookupUpkeepController) DeployUpkeeps(
+	t *testing.T,
+	backend *backends.SimulatedBackend,
+	owner *bind.TransactOpts,
+	count int,
+) error {
+	addresses := make([]common.Address, count)
+	contracts := make([]*log_triggered_streams_lookup_wrapper.LogTriggeredStreamsLookup, count)
+
+	// deploy n upkeep contracts
+	for x := 0; x < count; x++ {
+		addr, _, contract, err := log_triggered_streams_lookup_wrapper.DeployLogTriggeredStreamsLookup(
+			owner,
+			backend,
+			false,
+			false,
+		)
+
+		if err != nil {
+			require.NoError(t, err, "test dependent on contract deployment")
+
+			return err
+		}
+
+		addresses[x] = addr
+		contracts[x] = contract
+	}
+
+	backend.Commit()
+
+	c.count = count
+	c.addresses = addresses
+	c.contracts = contracts
+	c.contractsOwner = owner
+
+	return nil
+}
+
+func (c *feedLookupUpkeepController) RegisterAndFund(
+	t *testing.T,
+	f registerAndFundFunc,
```

Comment:
is it possible to move the `type registerAndFundFunc` and the function definition in `core/services/ocr2/plugins/ocr2keeper/integration_21_test.go` to here? So we can remove the additional layer?